helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/AssignableReplica.java

```diff
@@ -19,9 +19,121 @@
  * under the License.
  */
 
+import org.apache.helix.model.ResourceConfig;
+import org.apache.helix.model.StateModelDefinition;
+
+import java.io.IOException;
+import java.util.Map;
+
 /**
- * A placeholder before we have the implementation.
- *
  * This class represents a partition replication that needs to be allocated.
  */
-public class AssignableReplica { }
+public class AssignableReplica implements Comparable<AssignableReplica> {
+  private final String _partitionName;
+  private final String _resourceName;
+  private final String _resourceInstanceGroupTag;
+  private final int _resourceMaxPartitionsPerInstance;
+  private final Map<String, Integer> _capacityUsage;
+  // The priority of the replica's state
+  private final int _statePriority;
+  // The state of the replica
+  private final String _replicaState;
+
+  /**
+   * @param resourceConfig The resource config for the resource which contains the replication.
+   * @param partitionName  The replication's partition name.
+   * @param replicaState   The state of the replication.
+   * @param statePriority  The priority of the replication's state.
+   */
+  AssignableReplica(ResourceConfig resourceConfig, String partitionName, String replicaState,
+      int statePriority) {
+    _partitionName = partitionName;
+    _replicaState = replicaState;
+    _statePriority = statePriority;
+    _resourceName = resourceConfig.getResourceName();
+    _capacityUsage = fetchCapacityUsage(partitionName, resourceConfig);
```

Comment:
If you agree on the Builder pattern, here instead of making the `fetch` method private, we could publish it in the public method of builder. Something like

```
class Builder {
   public Builder capacityUsage(ResourceConfig resourceConfig, String partitionName) {
   
   }
}
```

It's Builder class' nature to do some heavy lifting works and public method just makes unit test so much easier.

Reply:
Even the constructor of Assignable Replica is not public. There is no need we make the construction methods public. For now, I think it is overdesigned to use builder pattern here.

We can revisit here if we are going to construct a replica object somewhere else.

Reply:
I would only advocate for the Builder pattern if there were more arguments (or if we expect to add more arguments to the constructor). To me, this is OK.